cmd/appengine/device_metadata.go

```diff
@@ -0,0 +1,158 @@
+var metadataCmd = &cobra.Command{
+	Use:   "metadata",
+	Short: "Interact with Device Metadata",
+	Long:  `Perform actions on Astarte Device Metadata.`,
+}
+
+var metadataListCmd = &cobra.Command{
+	Use:   "list <device_id_or_alias>",
+	Short: "List metadata",
+	Long: `List all metadata for a given Device.
+
+<device_id_or_alias> can be either a valid Astarte Device ID, or a Device Alias. In most cases,
+this is automatically determined - however, you can tweak this behavior by using --force-device-id or
+--force-id-type={device-id,alias}.`,
+	Example: `  astartectl appengine devices metadata list`,
+	Args:    cobra.ExactArgs(1),
+	RunE:    metadataListF,
+	Aliases: []string{"ls"},
+}
+
+var metadataSetCmd = &cobra.Command{
+	Use:   "set <device_id_or_alias> <key>=<value>",
+	Short: "Set device metadata",
+	Long: `Set a value to a metadata key in the given Device, in the form <key>=<value>.
+This is effectively an upsert operation: if the key already exist, the new value overwrites the old one.
+
+<device_id_or_alias> can be either a valid Astarte Device ID, or a Device Alias. In most cases,
+this is automatically determined - however, you can tweak this behavior by using --force-device-id or
+--force-id-type={device-id,alias}.`,
+	Example: `  astartectl appengine devices metadata add 2TBn-jNESuuHamE2Zo1anA room=kitchen`,
+	Args:    cobra.ExactArgs(2),
+	RunE:    metadataSetF,
+}
+
+var metadataRemoveCmd = &cobra.Command{
+	Use:   "remove <device_id_or_alias> <key>",
+	Short: "Remove metadata from a Device",
+	Long: `Remove metadata from the given Device, by specifying the key that has to be removed.
+
+<device_id_or_alias> can be either a valid Astarte Device ID, or a Device Alias. In most cases,
+this is automatically determined - however, you can tweak this behavior by using --force-device-id or
+--force-id-type={device-id,alias}.`,
+	Example: `  astartectl appengine devices metadata remove 2TBn-jNESuuHamE2Zo1anA room`,
+	Args:    cobra.ExactArgs(2),
+	RunE:    metadataRemoveF,
+	Aliases: []string{"rm"},
+}
+
+func init() {
+	devicesCmd.AddCommand(metadataCmd)
+
+	metadataListCmd.Flags().String("force-id-type", "", "When set, rather than autodetecting, it forces the device ID to be evaluated as a (device-id,alias).")
+	metadataSetCmd.Flags().String("force-id-type", "", "When set, rather than autodetecting, it forces the device ID to be evaluated as a (device-id,alias).")
+	metadataRemoveCmd.Flags().String("force-id-type", "", "When set, rather than autodetecting, it forces the device ID to be evaluated as a (device-id,alias).")
+
+	metadataCmd.AddCommand(
+		metadataListCmd,
+		metadataSetCmd,
+		metadataRemoveCmd,
+	)
+}
+
+func metadataListF(command *cobra.Command, args []string) error {
+	deviceID := args[0]
+	forceIDType, err := command.Flags().GetString("force-id-type")
+	if err != nil {
+		return err
+	}
+	deviceIdentifierType, err := deviceIdentifierTypeFromFlags(deviceID, forceIDType)
+	if err != nil {
+		return err
+	}
```

Comment:
This chunk is repeated in all the `metadata<Action>F` functions. Does it make sense isolating it as an helper function?

Reply:
This is already factored in `deviceIdentifierTypeFromFlags`. The only additional factoring would be passing `command` to that function instead of `forceIDType` and then extracting the flag - this could be legit (even if it would introduce the assumption that the flag has the same name everywhere, although that's reasonable), but given that this chunk of code is used in nearly every `appengine device` subcommand and not only in these new ones, it makes sense to possibily do that in a separate PR.